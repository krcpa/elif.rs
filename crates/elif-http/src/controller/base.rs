//! Service-Oriented Controller System
//! 
//! Provides a clean separation between HTTP handling (controllers) and business logic (services).
//! Controllers are thin HTTP handlers that delegate to injected services.

use std::{sync::Arc, pin::Pin, future::Future};
use crate::{
    request::{ElifState, ElifPath, ElifQuery},
    response::{ElifJson, ElifResponse},
};
use serde::{Serialize, Deserialize};
use serde_json::Value;

use elif_core::Container;
use crate::{HttpResult, response::ApiResponse};

/// Query parameters for pagination and filtering
#[derive(Debug, Serialize, Deserialize)]
pub struct QueryParams {
    pub page: Option<u32>,
    pub per_page: Option<u32>,
    pub sort: Option<String>,
    pub order: Option<String>,
    pub filter: Option<String>,
}

impl Default for QueryParams {
    fn default() -> Self {
        Self {
            page: Some(1),
            per_page: Some(20),
            sort: Some("id".to_string()),
            order: Some("asc".to_string()),
            filter: None,
        }
    }
}

/// Pagination metadata for API responses
#[derive(Debug, Serialize, Deserialize)]
pub struct PaginationMeta {
    pub page: u32,
    pub per_page: u32,
    pub total: Option<u64>,
    pub total_pages: Option<u32>,
    pub has_more: bool,
}

/// Base controller providing HTTP utilities (no infrastructure dependencies)
#[derive(Clone)]
pub struct BaseController;

impl BaseController {
    pub fn new() -> Self {
        Self
    }

    /// Validate and normalize pagination parameters
    pub fn normalize_pagination(&self, params: &QueryParams) -> (u32, u32, u64) {
        let page = params.page.unwrap_or(1).max(1);
        let per_page = params.per_page.unwrap_or(20).min(100).max(1);
        let offset = (page - 1) * per_page;
        (page, per_page, offset as u64)
    }

    /// Create standardized success response
    pub fn success_response<T: Serialize>(&self, data: T) -> HttpResult<ElifResponse> {
        let api_response = ApiResponse::success(data);
        Ok(ElifResponse::ok().json(&api_response)?)
    }

    /// Create standardized created response
    pub fn created_response<T: Serialize>(&self, data: T) -> HttpResult<ElifResponse> {
        let api_response = ApiResponse::success(data);
        Ok(ElifResponse::created().json(&api_response)?)
    }

    /// Create paginated response with metadata
    pub fn paginated_response<T: Serialize>(&self, data: Vec<T>, meta: PaginationMeta) -> HttpResult<ElifResponse> {
        let response_data = serde_json::json!({
            "data": data,
            "meta": meta
        });
        Ok(ElifResponse::ok().json(&response_data)?)
    }

    /// Create standardized delete response
    pub fn deleted_response<T: Serialize>(&self, resource_name: &str, deleted_id: Option<T>) -> HttpResult<ElifResponse> {
        let mut response_data = serde_json::json!({
            "message": format!("{} deleted successfully", resource_name)
        });
        
        if let Some(id) = deleted_id {
            response_data["deleted_id"] = serde_json::to_value(id)?;
        }
        
        let api_response = ApiResponse::success(response_data);
        Ok(ElifResponse::ok().json(&api_response)?)
    }
}

/// Send-safe controller trait for HTTP request handling
/// Controllers delegate business logic to injected services
pub trait Controller: Send + Sync {
    /// List resources with pagination
    fn index(
        &self,
        container: ElifState<Arc<Container>>,
        params: ElifQuery<QueryParams>,
    ) -> Pin<Box<dyn Future<Output = HttpResult<ElifResponse>> + Send>>;

    /// Get single resource by ID
    fn show(
        &self,
        container: ElifState<Arc<Container>>,
        id: ElifPath<String>,
    ) -> Pin<Box<dyn Future<Output = HttpResult<ElifResponse>> + Send>>;

    /// Create new resource
    fn create(
        &self,
        container: ElifState<Arc<Container>>,
        data: ElifJson<Value>,
    ) -> Pin<Box<dyn Future<Output = HttpResult<ElifResponse>> + Send>>;

    /// Update existing resource
    fn update(
        &self,
        container: ElifState<Arc<Container>>,
        id: ElifPath<String>,
        data: ElifJson<Value>,
    ) -> Pin<Box<dyn Future<Output = HttpResult<ElifResponse>> + Send>>;

    /// Delete resource
    fn destroy(
        &self,
        container: ElifState<Arc<Container>>,
        id: ElifPath<String>,
    ) -> Pin<Box<dyn Future<Output = HttpResult<ElifResponse>> + Send>>;
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[tokio::test]
    async fn test_base_controller_creation() {
        let _controller = BaseController::new();
    }

    #[tokio::test]
    async fn test_pagination_normalization() {
        let controller = BaseController::new();
        let params = QueryParams {
            page: Some(5),
            per_page: Some(10),
            ..Default::default()
        };
        
        let (page, per_page, offset) = controller.normalize_pagination(&params);
        assert_eq!(page, 5);
        assert_eq!(per_page, 10);
        assert_eq!(offset, 40);
    }

    #[tokio::test]
    async fn test_pagination_limits() {
        let controller = BaseController::new();
        let params = QueryParams {
            page: Some(0),
            per_page: Some(200),
            ..Default::default()
        };
        
        let (page, per_page, offset) = controller.normalize_pagination(&params);
        assert_eq!(page, 1);
        assert_eq!(per_page, 100);
        assert_eq!(offset, 0);
    }

    #[tokio::test]
    async fn test_success_response_creation() {
        let controller = BaseController::new();
        let data = json!({"message": "test"});
        let response = controller.success_response(data);
        assert!(response.is_ok());
    }

    #[tokio::test]
    async fn test_pagination_meta_creation() {
        let meta = PaginationMeta {
            page: 1,
            per_page: 20,
            total: Some(100),
            total_pages: Some(5),
            has_more: true,
        };
        
        assert_eq!(meta.page, 1);
        assert_eq!(meta.per_page, 20);
        assert_eq!(meta.total, Some(100));
    }
}