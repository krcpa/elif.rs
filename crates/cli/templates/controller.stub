use elif_http::prelude::*;
use elif_core::ServiceContainer;
use crate::models::{{ name | snake_case }}::{{ name | pascal_case }};
{% if validation %}
use crate::requests::{{ name | snake_case }}::{ Create{{ name | pascal_case }}Request, Update{{ name | pascal_case }}Request };
use crate::resources::{{ name | snake_case }}::{ {{ name | pascal_case }}Resource, {{ name | pascal_case }}Collection };
{% endif %}
use std::sync::Arc;

#[controller]
pub struct {{ name | pascal_case }}Controller {
    container: Arc<ServiceContainer>,
}

impl {{ name | pascal_case }}Controller {
    pub fn new(container: Arc<ServiceContainer>) -> Self {
        Self { container }
    }

    // <<<ELIF:BEGIN agent-editable:{{ name | snake_case }}-index>>>
    pub async fn index(&self, request: Request) -> Result<Response, HttpError> {
        let query = {{ name | pascal_case }}::query(){% if relationships %}{% for relationship in relationships %}{% if relationship.type == "belongs_to" %}
            .with("{{ relationship.related_model | snake_case }}"){% endif %}{% endfor %}{% endif %};
        
        let {{ name | snake_case | pluralize }} = query.paginate(request.per_page()).await
            .map_err(|e| HttpError::internal_server_error(format!("Database error: {}", e)))?;
        
        {% if validation %}
        Ok(Response::json({{ name | pascal_case }}Collection::new({{ name | snake_case | pluralize }})))
        {% else %}
        Ok(Response::json({{ name | snake_case | pluralize }}))
        {% endif %}
    }
    // <<<ELIF:END agent-editable:{{ name | snake_case }}-index>>>

    // <<<ELIF:BEGIN agent-editable:{{ name | snake_case }}-show>>>
    pub async fn show(&self, request: Request) -> Result<Response, HttpError> {
        let id = request.path_param("id")
            .map_err(|_| HttpError::bad_request("Invalid ID parameter"))?;
        
        let {{ name | snake_case }} = {{ name | pascal_case }}::find(&id).await
            .map_err(|e| HttpError::internal_server_error(format!("Database error: {}", e)))?
            .ok_or_else(|| HttpError::not_found("{{ name | pascal_case }} not found"))?;
        
        {% if validation %}
        Ok(Response::json({{ name | pascal_case }}Resource::new({{ name | snake_case }})))
        {% else %}
        Ok(Response::json({{ name | snake_case }}))
        {% endif %}
    }
    // <<<ELIF:END agent-editable:{{ name | snake_case }}-show>>>

    // <<<ELIF:BEGIN agent-editable:{{ name | snake_case }}-store>>>
    pub async fn store(&self, mut request: Request) -> Result<Response, HttpError> {
        {% if auth %}
        let user = request.require_user()
            .map_err(|_| HttpError::unauthorized("Authentication required"))?;
        
        {% if policy %}
        user.can("create", {{ name | pascal_case }}::class())
            .map_err(|_| HttpError::forbidden("Not authorized to create {{ name | pluralize | lower }}"))?;
        {% endif %}
        {% endif %}
        
        {% if validation %}
        let data: Create{{ name | pascal_case }}Request = request.validate_json()
            .map_err(|e| HttpError::unprocessable_entity(format!("Validation error: {}", e)))?;
        {% else %}
        let data: {{ name | pascal_case }} = request.json().await
            .map_err(|e| HttpError::bad_request(format!("Invalid JSON: {}", e)))?;
        {% endif %}
        
        let {{ name | snake_case }} = {{ name | pascal_case }} {
            {% for field in fields %}
            {% if not field.pk %}
            {{ field.name | snake_case }}: {% if validation %}data.{{ field.name | snake_case }}{% else %}data.{{ field.name | snake_case }}{% endif %},
            {% endif %}
            {% endfor %}
            {% if timestamps %}
            created_at: chrono::Utc::now(),
            updated_at: chrono::Utc::now(),
            {% endif %}
            {% if soft_delete %}
            deleted_at: None,
            {% endif %}
        };

        let saved_{{ name | snake_case }} = {{ name | snake_case }}.save().await
            .map_err(|e| HttpError::internal_server_error(format!("Database error: {}", e)))?;
        
        {% if validation %}
        Ok(Response::json({{ name | pascal_case }}Resource::new(saved_{{ name | snake_case }})).status(201))
        {% else %}
        Ok(Response::json(saved_{{ name | snake_case }}).status(201))
        {% endif %}
    }
    // <<<ELIF:END agent-editable:{{ name | snake_case }}-store>>>

    // <<<ELIF:BEGIN agent-editable:{{ name | snake_case }}-update>>>
    pub async fn update(&self, mut request: Request) -> Result<Response, HttpError> {
        let id = request.path_param("id")
            .map_err(|_| HttpError::bad_request("Invalid ID parameter"))?;
        
        {% if auth %}
        let user = request.require_user()
            .map_err(|_| HttpError::unauthorized("Authentication required"))?;
        {% endif %}
        
        let mut {{ name | snake_case }} = {{ name | pascal_case }}::find(&id).await
            .map_err(|e| HttpError::internal_server_error(format!("Database error: {}", e)))?
            .ok_or_else(|| HttpError::not_found("{{ name | pascal_case }} not found"))?;
        
        {% if policy %}
        user.can("update", &{{ name | snake_case }})
            .map_err(|_| HttpError::forbidden("Not authorized to update this {{ name | lower }}"))?;
        {% endif %}
        
        {% if validation %}
        let data: Update{{ name | pascal_case }}Request = request.validate_json()
            .map_err(|e| HttpError::unprocessable_entity(format!("Validation error: {}", e)))?;
        {% else %}
        let data: {{ name | pascal_case }} = request.json().await
            .map_err(|e| HttpError::bad_request(format!("Invalid JSON: {}", e)))?;
        {% endif %}
        
        // Update fields
        {% for field in fields %}
        {% if not field.pk and field.name != "created_at" %}
        {{ name | snake_case }}.{{ field.name | snake_case }} = {% if validation %}data.{{ field.name | snake_case }}{% else %}data.{{ field.name | snake_case }}{% endif %};
        {% endif %}
        {% endfor %}
        {% if timestamps %}
        {{ name | snake_case }}.updated_at = chrono::Utc::now();
        {% endif %}

        let updated_{{ name | snake_case }} = {{ name | snake_case }}.save().await
            .map_err(|e| HttpError::internal_server_error(format!("Database error: {}", e)))?;
        
        {% if validation %}
        Ok(Response::json({{ name | pascal_case }}Resource::new(updated_{{ name | snake_case }})))
        {% else %}
        Ok(Response::json(updated_{{ name | snake_case }}))
        {% endif %}
    }
    // <<<ELIF:END agent-editable:{{ name | snake_case }}-update>>>

    // <<<ELIF:BEGIN agent-editable:{{ name | snake_case }}-destroy>>>
    pub async fn destroy(&self, request: Request) -> Result<Response, HttpError> {
        let id = request.path_param("id")
            .map_err(|_| HttpError::bad_request("Invalid ID parameter"))?;
        
        {% if auth %}
        let user = request.require_user()
            .map_err(|_| HttpError::unauthorized("Authentication required"))?;
        {% endif %}
        
        let {{ name | snake_case }} = {{ name | pascal_case }}::find(&id).await
            .map_err(|e| HttpError::internal_server_error(format!("Database error: {}", e)))?
            .ok_or_else(|| HttpError::not_found("{{ name | pascal_case }} not found"))?;
        
        {% if policy %}
        user.can("delete", &{{ name | snake_case }})
            .map_err(|_| HttpError::forbidden("Not authorized to delete this {{ name | lower }}"))?;
        {% endif %}
        
        {% if soft_delete %}
        {{ name | snake_case }}.soft_delete().await
        {% else %}
        {{ name | snake_case }}.delete().await
        {% endif %}
            .map_err(|e| HttpError::internal_server_error(format!("Database error: {}", e)))?;
        
        Ok(Response::no_content())
    }
    // <<<ELIF:END agent-editable:{{ name | snake_case }}-destroy>>>
}