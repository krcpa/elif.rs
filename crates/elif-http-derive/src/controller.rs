//! Controller macro implementation
//!
//! Provides the #[controller] macro for defining controller base path and metadata.

use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, ImplItem, ItemImpl, ItemStruct, LitStr};

use crate::utils::{
    extract_http_method_info, extract_middleware_from_attrs, extract_param_types_from_attrs,
    extract_path_parameters,
};

/// Controller macro for defining controller base path and metadata
///
/// This macro should be applied to impl blocks to enable route registration.
///
/// Example:
/// ```rust,ignore
/// pub struct UserController;
///
/// #[controller("/users")]
/// impl UserController {
///     #[get("/{id}")]
///     async fn show(&self, req: ElifRequest) -> HttpResult<ElifResponse> {
///         // handler implementation
///     }
/// }
/// ```
pub fn controller_impl(args: TokenStream, input: TokenStream) -> TokenStream {
    let path_lit = parse_macro_input!(args as LitStr);
    let base_path = path_lit.value();

    // Try to parse as impl block first (new approach)
    if let Ok(mut input_impl) = syn::parse::<ItemImpl>(input.clone()) {
        let self_ty = &input_impl.self_ty;
        let struct_name = if let syn::Type::Path(type_path) = &**self_ty {
            if let Some(segment) = type_path.path.segments.last() {
                segment.ident.to_string()
            } else {
                return syn::Error::new_spanned(
                    self_ty, 
                    "Cannot extract struct name from type path. Hint: Use a simple struct name like `MyController`."
                )
                .to_compile_error()
                .into();
            }
        } else {
            return syn::Error::new_spanned(
                self_ty, 
                "Expected a simple type for impl block. Hint: Apply #[controller] to `impl MyStruct { ... }` not complex types."
            )
            .to_compile_error()
            .into();
        };

        // Collect route information from methods
        let mut routes = Vec::new();
        let mut method_handlers = Vec::new();

        for item in &input_impl.items {
            if let ImplItem::Fn(method) = item {
                let method_name = &method.sig.ident;

                // Check for HTTP method attributes
                if let Some((http_method, path)) = extract_http_method_info(&method.attrs) {
                    let handler_name = method_name.to_string();
                    let handler_name_lit = LitStr::new(&handler_name, method_name.span());

                    // Convert http_method ident to proper HttpMethod enum variant
                    let http_method_variant = match http_method.to_string().as_str() {
                        "get" => quote! { GET },
                        "post" => quote! { POST },
                        "put" => quote! { PUT },
                        "delete" => quote! { DELETE },
                        "patch" => quote! { PATCH },
                        "head" => quote! { HEAD },
                        "options" => quote! { OPTIONS },
                        _ => unreachable!(
                            "extract_http_method_info should only return valid HTTP methods"
                        ),
                    };

                    // Extract middleware from method attributes
                    let middleware = extract_middleware_from_attrs(&method.attrs);
                    let middleware_vec = quote! { vec![#(#middleware.to_string()),*] };

                    // Extract path parameters from the route path
                    let path_params = extract_path_parameters(&path);

                    // Extract parameter type specifications from #[param] attributes
                    let param_types = extract_param_types_from_attrs(&method.attrs);

                    // Build parameter metadata with proper types
                    let mut param_tokens = Vec::new();
                    for param_name in &path_params {
                        // Get the type from #[param] attributes, default to String
                        let param_type = param_types
                            .get(param_name)
                            .cloned()
                            .unwrap_or_else(|| "String".to_string());
                        let param_type_enum = match param_type.as_str() {
                            "String" => quote! { ParamType::String },
                            "Integer" => quote! { ParamType::Integer },
                            "Uuid" => quote! { ParamType::Uuid },
                            _ => quote! { ParamType::String }, // Default fallback
                        };

                        param_tokens.push(quote! {
                            RouteParam::new(#param_name, #param_type_enum)
                        });
                    }

                    routes.push(quote! {
                        ControllerRoute {
                            method: HttpMethod::#http_method_variant,
                            path: #path.to_string(),
                            handler_name: #handler_name.to_string(),
                            middleware: #middleware_vec,
                            params: vec![#(#param_tokens),*],
                        }
                    });

                    // Generate handler for async dispatch with Arc<Self>
                    method_handlers.push(quote! {
                        #handler_name_lit => {
                            self.#method_name(request).await
                        }
                    });
                }
            }
        }

        // Add constants to the impl block
        input_impl.items.push(syn::parse_quote! {
            pub const BASE_PATH: &'static str = #base_path;
        });

        input_impl.items.push(syn::parse_quote! {
            pub const CONTROLLER_NAME: &'static str = #struct_name;
        });

        // Generate method handlers for async dispatch  
        let method_match_arms = &method_handlers;

        // Check if this controller needs dependency injection and extract constructor info
        let mut needs_dependency_injection = false;
        let mut constructor_param_types = Vec::new();
        let mut constructor_param_names = Vec::new();
        
        for item in &input_impl.items {
            if let syn::ImplItem::Fn(method) = item {
                if method.sig.ident == "new" && !method.sig.inputs.is_empty() {
                    needs_dependency_injection = true;
                    // Extract parameter types and names from constructor
                    // Note: new() is a static method, so no 'self' parameter to skip
                    for (i, input) in method.sig.inputs.iter().enumerate() {
                        if let syn::FnArg::Typed(pat_type) = input {
                            if let syn::Type::Path(type_path) = &*pat_type.ty {
                                if let Some(segment) = type_path.path.segments.last() {
                                    constructor_param_types.push(segment.ident.clone());
                                    // Generate parameter variable names like param_0, param_1, etc.
                                    let param_name = syn::Ident::new(&format!("param_{}", i), segment.ident.span());
                                    constructor_param_names.push(param_name);
                                }
                            }
                        }
                    }
                    break;
                }
            }
        }

        // Generate appropriate trait implementations based on DI needs
        let ioc_controllable_impl = if needs_dependency_injection {
            quote! {
                // Auto-generated IocControllable implementation for dependency injection
                impl ::elif_http::controller::factory::IocControllable for #self_ty {
                    fn from_ioc_container(
                        container: &::elif_core::container::IocContainer,
                        _scope: Option<&::elif_core::container::ScopeId>,
                    ) -> Result<Self, String> {
                        // Auto-resolve dependencies from container
                        // This provides automatic dependency injection for common patterns
                        Self::from_container_auto(container)
                    }
                }

                impl #self_ty {
                    /// Auto-generated dependency resolution method
                    /// This attempts to auto-resolve dependencies using Default implementations
                    /// Controllers can override this for custom dependency injection logic
                    fn from_container_auto(_container: &::elif_core::container::IocContainer) -> Result<Self, String> {
                        // Attempt to create the controller using Default implementations of dependencies
                        // This works for services that implement Default trait
                        match Self::try_new_with_defaults() {
                            Ok(controller) => Ok(controller),
                            Err(e) => Err(format!(
                                "Controller {} requires dependency injection. {}\n\
                                Please either:\n\
                                1. Ensure all dependencies implement Default trait, or\n\
                                2. Implement a custom `from_container_auto` method, or\n\
                                3. Use router.controller_from_container::<{}>()\n\
                                4. Register dependencies in IoC container and use Injectable trait",
                                stringify!(#self_ty),
                                e,
                                stringify!(#self_ty)
                            ))
                        }
                    }

                    /// Try to create controller with Default implementations of dependencies
                    fn try_new_with_defaults() -> Result<Self, String> {
                        // This is a fallback - try to call new() expecting Default dependencies
                        // If this fails, it means dependencies don't implement Default
                        Self::try_new_auto()
                    }

                    /// Auto-generated attempt to create controller with Default dependencies
                    fn try_new_auto() -> Result<Self, String> {
                        // This will work if all constructor parameters implement Default
                        Self::new_with_default_deps()
                    }

                    /// Template method for creating with default dependencies
                    /// This gets specialized per controller based on constructor signature
                    fn new_with_default_deps() -> Result<Self, String> {
                        // Generate dependency instances using Default trait
                        #(
                            let #constructor_param_names = #constructor_param_types::default();
                        )*
                        Ok(Self::new(#(#constructor_param_names),*))
                    }
                }
            }
        } else {
            // For controllers without dependencies, provide empty implementation block
            quote! {
                // No additional trait implementations needed for parameterless controllers
                // They should implement Default themselves or use #[derive(Default)]
            }
        };



        let registration_code = if needs_dependency_injection {
            quote! {
                // For IoC controllers, provide helpful compile-time guidance
                const _: () = {
                    // This is a marker to indicate this controller needs IoC container registration
                    // To register: router.controller_from_container::<ControllerType>()
                };
            }
        } else {
            quote! {
                // For simple controllers, use traditional auto-registration
                ::elif_http::__controller_auto_register! {
                    #struct_name,
                    #self_ty
                }
            }
        };

        let expanded = quote! {
            #input_impl

            #[::async_trait::async_trait]
            impl ElifController for #self_ty {
                fn name(&self) -> &str {
                    #struct_name
                }

                fn base_path(&self) -> &str {
                    #base_path
                }

                fn routes(&self) -> Vec<ControllerRoute> {
                    vec![
                        #(#routes),*
                    ]
                }

                async fn handle_request(
                    self: std::sync::Arc<Self>,
                    method_name: String,
                    request: ElifRequest,
                ) -> HttpResult<ElifResponse> {
                    match method_name.as_str() {
                        #(#method_match_arms,)*
                        _ => {
                            Ok(ElifResponse::not_found()
                                .text(&format!("Handler '{}' not found", method_name)))
                        }
                    }
                }
            }

            #ioc_controllable_impl

            // Registration code (conditional based on DI needs)
            #registration_code
        };

        TokenStream::from(expanded)
    } else if let Ok(input_struct) = syn::parse::<ItemStruct>(input) {
        // Legacy support: If applied to struct, just add constants
        let struct_name = &input_struct.ident;
        let struct_name_str = struct_name.to_string();

        let expanded = quote! {
            #input_struct

            impl #struct_name {
                pub const BASE_PATH: &'static str = #base_path;
                pub const CONTROLLER_NAME: &'static str = #struct_name_str;
            }
        };

        TokenStream::from(expanded)
    } else {
        syn::Error::new(
            proc_macro2::Span::call_site(),
            "controller attribute must be applied to an impl block or struct. Hint: Use `#[controller(\"/path\")] impl MyController { ... }` or `#[controller(\"/path\")] struct MyController;`"
        )
        .to_compile_error()
        .into()
    }
}
