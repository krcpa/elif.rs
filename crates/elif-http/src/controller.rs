//! Service-Oriented Controller System
//! 
//! Provides a clean separation between HTTP handling (controllers) and business logic (services).
//! Controllers are thin HTTP handlers that delegate to injected services.

use std::{sync::Arc, pin::Pin, future::Future};
use axum::{
    extract::{State, Path, Query},
    response::{Json, Response, IntoResponse},
    http::StatusCode,
};
use serde::{Serialize, Deserialize};
use serde_json::Value;

use elif_core::Container;
use crate::{HttpResult, HttpError, ApiResponse};

/// Query parameters for pagination and filtering
#[derive(Debug, Serialize, Deserialize)]
pub struct QueryParams {
    pub page: Option<u32>,
    pub per_page: Option<u32>,
    pub sort: Option<String>,
    pub order: Option<String>,
    pub filter: Option<String>,
}

impl Default for QueryParams {
    fn default() -> Self {
        Self {
            page: Some(1),
            per_page: Some(20),
            sort: Some("id".to_string()),
            order: Some("asc".to_string()),
            filter: None,
        }
    }
}

/// Pagination metadata for API responses
#[derive(Debug, Serialize, Deserialize)]
pub struct PaginationMeta {
    pub page: u32,
    pub per_page: u32,
    pub total: Option<u64>,
    pub total_pages: Option<u32>,
    pub has_more: bool,
}

/// Base controller providing HTTP utilities (no infrastructure dependencies)
#[derive(Clone)]
pub struct BaseController;

impl BaseController {
    pub fn new() -> Self {
        Self
    }

    /// Validate and normalize pagination parameters
    pub fn normalize_pagination(&self, params: &QueryParams) -> (u32, u32, u64) {
        let page = params.page.unwrap_or(1).max(1);
        let per_page = params.per_page.unwrap_or(20).min(100).max(1);
        let offset = (page - 1) * per_page;
        (page, per_page, offset as u64)
    }

    /// Create standardized success response
    pub fn success_response<T: Serialize>(&self, data: T) -> HttpResult<Response> {
        let api_response = ApiResponse::success(data);
        Ok((StatusCode::OK, Json(api_response)).into_response())
    }

    /// Create standardized created response
    pub fn created_response<T: Serialize>(&self, data: T) -> HttpResult<Response> {
        let api_response = ApiResponse::success(data);
        Ok((StatusCode::CREATED, Json(api_response)).into_response())
    }

    /// Create paginated response with metadata
    pub fn paginated_response<T: Serialize>(&self, data: Vec<T>, meta: PaginationMeta) -> HttpResult<Response> {
        let response_data = serde_json::json!({
            "data": data,
            "meta": meta
        });
        let api_response = ApiResponse::success(response_data);
        Ok((StatusCode::OK, Json(api_response)).into_response())
    }

    /// Create standardized delete response
    pub fn deleted_response<T: Serialize>(&self, resource_name: &str, deleted_id: Option<T>) -> HttpResult<Response> {
        let mut response_data = serde_json::json!({
            "message": format!("{} deleted successfully", resource_name)
        });
        
        if let Some(id) = deleted_id {
            response_data["deleted_id"] = serde_json::to_value(id)?;
        }
        
        let api_response = ApiResponse::success(response_data);
        Ok((StatusCode::OK, Json(api_response)).into_response())
    }
}

/// Send-safe controller trait for HTTP request handling
/// Controllers delegate business logic to injected services
pub trait Controller: Send + Sync {
    /// List resources with pagination
    fn index(
        &self,
        container: State<Arc<Container>>,
        params: Query<QueryParams>,
    ) -> Pin<Box<dyn Future<Output = HttpResult<Response>> + Send>>;

    /// Get single resource by ID
    fn show(
        &self,
        container: State<Arc<Container>>,
        id: Path<String>,
    ) -> Pin<Box<dyn Future<Output = HttpResult<Response>> + Send>>;

    /// Create new resource
    fn create(
        &self,
        container: State<Arc<Container>>,
        data: Json<Value>,
    ) -> Pin<Box<dyn Future<Output = HttpResult<Response>> + Send>>;

    /// Update existing resource
    fn update(
        &self,
        container: State<Arc<Container>>,
        id: Path<String>,
        data: Json<Value>,
    ) -> Pin<Box<dyn Future<Output = HttpResult<Response>> + Send>>;

    /// Delete resource
    fn destroy(
        &self,
        container: State<Arc<Container>>,
        id: Path<String>,
    ) -> Pin<Box<dyn Future<Output = HttpResult<Response>> + Send>>;
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[tokio::test]
    async fn test_base_controller_creation() {
        let _controller = BaseController::new();
    }

    #[tokio::test]
    async fn test_pagination_normalization() {
        let controller = BaseController::new();
        let params = QueryParams {
            page: Some(5),
            per_page: Some(10),
            ..Default::default()
        };
        
        let (page, per_page, offset) = controller.normalize_pagination(&params);
        assert_eq!(page, 5);
        assert_eq!(per_page, 10);
        assert_eq!(offset, 40);
    }

    #[tokio::test]
    async fn test_pagination_limits() {
        let controller = BaseController::new();
        let params = QueryParams {
            page: Some(0),
            per_page: Some(200),
            ..Default::default()
        };
        
        let (page, per_page, offset) = controller.normalize_pagination(&params);
        assert_eq!(page, 1);
        assert_eq!(per_page, 100);
        assert_eq!(offset, 0);
    }

    #[tokio::test]
    async fn test_success_response_creation() {
        let controller = BaseController::new();
        let data = json!({"message": "test"});
        let response = controller.success_response(data);
        assert!(response.is_ok());
    }

    #[tokio::test]
    async fn test_pagination_meta_creation() {
        let meta = PaginationMeta {
            page: 1,
            per_page: 20,
            total: Some(100),
            total_pages: Some(5),
            has_more: true,
        };
        
        assert_eq!(meta.page, 1);
        assert_eq!(meta.per_page, 20);
        assert_eq!(meta.total, Some(100));
    }
}