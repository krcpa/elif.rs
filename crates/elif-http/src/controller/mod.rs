pub mod base;
pub mod pagination;

pub use base::{
    Controller, BaseController, 
    ElifController, ControllerRoute, RouteParam
};
pub use pagination::{QueryParams, PaginationMeta};