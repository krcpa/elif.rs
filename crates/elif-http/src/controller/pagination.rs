//! Pagination utilities for controllers

use serde::{Deserialize, Serialize};

/// Query parameters for pagination and filtering
#[derive(Debug, Serialize, Deserialize)]
pub struct QueryParams {
    pub page: Option<u32>,
    pub per_page: Option<u32>,
    pub search: Option<String>,
    pub filter: Option<String>,
    pub sort: Option<String>,
    pub order: Option<String>,
}

impl Default for QueryParams {
    fn default() -> Self {
        Self {
            page: Some(1),
            per_page: Some(10),
            search: None,
            filter: None,
            sort: None,
            order: None,
        }
    }
}

/// Pagination metadata for responses
#[derive(Debug, Serialize)]
pub struct PaginationMeta {
    pub current_page: u32,
    pub per_page: u32,
    pub total_items: u64,
    pub total_pages: u32,
    pub has_next: bool,
    pub has_prev: bool,
}

impl PaginationMeta {
    pub fn new(page: u32, per_page: u32, total: u64) -> Self {
        let total_pages = if per_page > 0 {
            (total as f64 / per_page as f64).ceil() as u32
        } else {
            0
        };

        Self {
            current_page: page,
            per_page,
            total_items: total,
            total_pages,
            has_next: page < total_pages,
            has_prev: page > 1,
        }
    }
}
