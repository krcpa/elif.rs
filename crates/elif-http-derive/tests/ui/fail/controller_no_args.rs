//! Test that controller macro without arguments fails

use elif_http_derive::controller;

// This should fail - controller macro requires a path argument
#[controller]
pub struct BadController;

fn main() {}