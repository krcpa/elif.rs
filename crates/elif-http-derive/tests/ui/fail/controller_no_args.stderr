error: unexpected end of input, expected string literal
 --> tests/ui/fail/controller_no_args.rs:6:1
  |
6 | #[controller]
  | ^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `controller` (in Nightly builds, run with -Z macro-backtrace for more info)
